Show stream subscriptions and user groups in profile modal.

m-e-l-u-h-a-n opened the pull request:
Fixes: #16233 .

**chat.zulip.org thread:** [https://chat.zulip.org/#narrow/stream/6-frontend/topic/add.20streams.20and.20user.20groups.20of.20an.20user.20in.20full.20profile.20mo.2E.2E.2E](https://chat.zulip.org/#narrow/stream/6-frontend/topic/add.20streams.20and.20user.20groups.20of.20an.20user.20in.20full.20profile.20mo.2E.2E.2E)

This completes work that was started by @Abhirup-99 in #16663.


@shubhamdhama  @pragatiagrawal31 FYI.


**GIFs or screenshots:** 


![full_profile_view](https://user-images.githubusercontent.com/63504956/120496080-e65be100-c3da-11eb-9d03-4f77c6ae18d9.gif)

timabbott commented:
Looking at this UI:

![image](https://user-images.githubusercontent.com/2746074/120580465-1bb5fe00-c3de-11eb-8ccd-51ed4311e31b.png)

My proposal is that we only show the avatar and content to the right of it when you're looking at the "Profile" tab -- so that the "Streams" and "User groups" tabs (we should use those terms) just have the content about the user's streams/groups, and not some big block above it (which feels distracting).

timabbott commented on static/templates/user_profile_modal.hbs:
```diff
@@ -67,23 +67,19 @@
-                        {{/each}}
-                    </ul>
+                <div class="tabcontent" id="streams">
+                    <table class="user-stream-list" data-empty="No stream subscriptions."></table>
```
These `data-empty` values should always be tagged for translation.

m-e-l-u-h-a-n replied:
Updated it.

timabbott commented on static/js/popovers.js:
```diff
@@ -404,6 +412,19 @@ function render_user_stream_list(streams, user) {
 
+function render_user_group_list(groups, user) {
+    groups.sort(compare_by_name);
+    const container = $("#groups .user-group-list");
```
`#groups` seems like a really general selector that's asking for bugs -- the container element should be found via `#user-profile-modal`, and we should use the `#groups` ID for this widget.

m-e-l-u-h-a-n replied:
Hmm, this id is also used in toggling the tabs, so I did not remove the id but renamed it to `groups-tab`. And for container updated it so that we are now selecting it via `#user-profile-modal .user-group-list`.

timabbott commented:
I merged the first commit as a46cb14845468db77604d7fe2dd58fe5f2e72594, thanks @m-e-l-u-h-a-n!

Posted comments on the rest; it's possible you just want to squash them, fix the issues, and add a `Co-Author-By` entry in the commit message so that @Abhirup-99 is credited as well. (I may have a typo in the name; it's a standard Git thing and `git log` has some examples).

m-e-l-u-h-a-n commented:
I have updated this pr, by making following changes:
* Generic profile information is only shown in `Profile` tab, `Streams` and `User groups` tab show only streams and user groups lists respectively.
* Order of tabs is now `Profile, Streams, User groups` earlier it was `Profile, User groups, Streams`.
* Container for list widget is selected as `#user-profile-modal .user-group-list` and `#user-profile-modal .user-stream-list` for groups and streams respectively.
* Squashed the commits.

It is ready for another look.

m-e-l-u-h-a-n commented:
@timabbott this has been in completed state for a while. I've updated it to clear the conflicts, can you have a look at it.

timabbott commented:
@alya can you take a look at the visuals here?  

![full_profile_view](https://user-images.githubusercontent.com/63504956/120616589-004b0180-c477-11eb-96d6-2a1e82b8380d.gif)

timabbott commented on static/js/popovers.js:
```diff
@@ -372,6 +377,54 @@ export function hide_user_profile() {
 
+function compare_by_name(a, b) {
+    return a.name.localeCompare(b.name);
+}
```
We should use `util.strcmp` here; see the comments on `localeCompare` being really slow for why.  

Would you be up for doing a sweep?  It looks like `git grep localeCompare` finds a number of existing uses.

timabbott commented on static/styles/popovers.css:
```diff
@@ -351,6 +358,59 @@ ul {
         }
     }
+
+    .tab-data {
```
I tweaked this to add a 20px bottom margin to this container, since it looked bad when scrolled to the bottom.

timabbott commented:
I think this is a point I'd like to iterate forward from.  I'm not happy with the visuals for the new tabs, but it's not particularly worse than the modal as-is, so it seems worth doing that as follow-up work.  I merged this with the change above with a bonus commit fixing a related regression I had introduced since your last comment.

@m-e-l-u-h-a-n for follow-ups:
* We should do a sweep to change the `strcmp` thing noted above.
* We should add an "unsubscribe" button to those tables if the current user is an administrator with permission to unsubscribe the user.
* We should think about adding a "Filter" widget.
* Can you start a thread on #design asking for ideas on how to improve the visuals of this modal?  I'm not sure the bordered table styling is helping us, but the modal in general needs work on its visuals.

abhijeetbodas2001 commented on static/js/popovers.js:
```diff
@@ -380,7 +433,8 @@ export function show_user_profile(user) {
         .filter((f) => f.name !== undefined);
-
+    const user_streams = stream_data.subscribed_subs();
+    const groups_of_user = user_groups.get_user_groups_of_user(user.user_id);
```
It seems `subscribed_subs` returns the streams the current user has subscribed to? (As opposed to that of `user.user_id`?)

m-e-l-u-h-a-n replied:
Hmm this is the reason I was seeing same stream list for almost all profile modals. Thanks for reporting, I'll fix it, actually it was a part of commits from abhirup's PR which I did not changed(assuming that it was reviewed) .

timabbott replied:
Good find.

Generally, you shouldn't assume that code in a previous PR is correct or reviewed; it's always best to understand all of the code yourself when picking up an existing project.
